js/store.js

```diff
@@ -296,58 +300,64 @@ const store = {
     // and the third party is visible (i.e. not whitelisted)
-    if (!('thirdPartyHostnames' in  firstParty)
-      || !firstParty['thirdPartyHostnames'].includes(target)) {
-      if (!('isVisible' in thirdParty) || !thirdParty['isVisible']) {
-        if (!(this.onWhiteList(origin, target))) {
-          // show third party
-          thirdParty['isVisible'] = true;
-          isNewThirdParty = true;
-        } else {
+    if (!this.isFirstPartyConnectedToThirdParty(firstParty, target)) {
+      if (!this.isVisibleThirdParty(thirdParty)) {
+        if (this.onWhiteList(origin, target)) {
           // hide third party
           thirdParty['isVisible'] = false;
+        } else {
+          // show third party and all its connections
+          thirdParty['isVisible'] = true;
+          isNewThirdParty = true;
+          thirdParty['firstPartyHostnames'].forEach((firstPartyHostname) => {
+            this.addFirstPartyLink(firstPartyHostname, target);
+          });
+          shouldUpdate = true;
         }
       }
-      if (thirdParty['isVisible'] && isNewFirstParty) {
+      if (this.isVisibleThirdParty(thirdParty) && !isNewThirdParty) {
         // an existing visible third party links to a new first party
-        thirdParty['firstPartyHostnames'].forEach((firstPartyHostname) => {
-          this.addFirstPartyLink(firstPartyHostname, target);
-        });
+        this.addFirstPartyLink(origin, target);
+        shouldUpdate = true;
       }
     }
 
-    // merge data with thirdParty website object
-    for (const key in data) {
-      thirdParty[key] = data[key];
-    }
-
-    this.setWebsite(target, thirdParty);
-
-    // should update if:
-    //   - there is a new third party, OR
-    //   - there is a new link between an old third party and a new first party
-    //   AND
-    //   - the third party is visible (i.e. not whitelisted)
-    const shouldUpdate = (isNewThirdParty || isNewFirstParty)
-      && thirdParty['isVisible'];
+    this.setWebsite(target, thirdParty, data);
 
     if (shouldUpdate) {
       this.updateChild(this.outputWebsite(target, thirdParty));
     }
   },
 
+  isFirstPartyConnectedToThirdParty(firstParty, thirdPartyHostname) {
+    if (!('thirdPartyHostnames' in  firstParty)
+      || !firstParty['thirdPartyHostnames'].includes(thirdPartyHostname)) {
+      return false;
+    }
+    return true;
+  },
+
+  isVisibleThirdParty(thirdParty) {
+    if (!('isVisible' in thirdParty) || !thirdParty['isVisible']) {
+      return false;
+    }
+    return true;
+  },
+
   addFirstPartyLink(firstPartyHostname, thirdPartyHostname) {
     const firstParty = this.getWebsite(firstPartyHostname);
     if (!('thirdPartyHostnames' in firstParty)) {
       firstParty['thirdPartyHostnames'] = [];
     }
-    firstParty['thirdPartyHostnames'].push(thirdPartyHostname);
-    this.setFirstParty(firstPartyHostname, firstParty);
+    if (!this
+      .isFirstPartyConnectedToThirdParty(firstParty, thirdPartyHostname)) {
```

Comment:
Okay yeah this was due to the 80 character line length limit in eslint. I shortened the name of this method slightly to `isFirstPartyLinkedToThirdParty`. :)